src/jdk.jdwp.agent/share/native/libjdwp/util.c

```diff
@@ -1099,31 +1106,394 @@ debugMonitorNotifyAll(jrawMonitorID monitor)
+        dbgRawMonitor_unlock();
+    }
+
+    dbg_monitor->entryCount++;
+    JNI_FUNC_PTR(env,DeleteLocalRef)(env, current_thread);
+}
+
+void
+debugMonitorExit(DebugRawMonitor *dbg_monitor)
+{
+    jrawMonitorID monitor = dbg_monitor->monitor;
+
+    if (!gdata->rankedMonitors) {
+        debugMonitorExit_norank(monitor);
+        return;
+    }
+
+    JNIEnv *env = getEnv();
+    jthread current_thread = threadControl_currentThread();
+
+    if (gdata->vmDead && current_thread == NULL) {
+        return;
+    }
+    JNI_FUNC_PTR(env,DeleteLocalRef)(env, current_thread);
+
+    JDI_ASSERT(dbg_monitor->entryCount > 0);
+    // Assert that the current thread owns this monitor.
+    assertIsCurrentThread(env, dbg_monitor->ownerThread);
+
+    dbg_monitor->entryCount--;
+    if (dbg_monitor->entryCount == 0) {
+        // Release ownership of the DebugRawMonitor before RawMonitorExit actually exits it.
+        dbgRawMonitor_lock();
+        tossGlobalRef(env, (jobject*)&(dbg_monitor->ownerThread));
+        dbgRawMonitor_unlock();
+
+        debugMonitorExit_norank(monitor);
+    }
+}
+
+void
+debugMonitorWait(DebugRawMonitor *dbg_monitor)
+{
+    jrawMonitorID monitor = dbg_monitor->monitor;
+
+    if (!gdata->rankedMonitors) {
+        debugMonitorWait_norank(monitor);
+        return;
+    }
+
+    jthread thread = dbg_monitor->ownerThread;
+    int entryCount = dbg_monitor->entryCount;
+    jthread current_thread = threadControl_currentThread();
+
+    if (gdata->vmDead && current_thread == NULL) {
+        return;
+    }
+    JNIEnv *env = getEnv();
+    JNI_FUNC_PTR(env,DeleteLocalRef)(env, current_thread);
+
+    JDI_ASSERT(thread != NULL);
+    // Assert that the current thread owns this monitor.
+    assertIsCurrentThread(env, dbg_monitor->ownerThread);
+
+    // Release ownership of the DebugRawMonitor before RawMonitorWait actually exits it.
+    dbgRawMonitor_lock();
+    dbg_monitor->ownerThread = NULL;
+    dbg_monitor->entryCount = 0;
+    dbgRawMonitor_unlock();
+
+    debugMonitorWait_norank(monitor);
+
+    // Now that we have re-entered the monitor, reclaim ownership of the DebugRawMonitor
+    // by saving the thread into it and restoring the entry count.
+    dbgRawMonitor_lock();
+    dbg_monitor->ownerThread = thread;
+    dbg_monitor->entryCount = entryCount;
+    dbgRawMonitor_unlock();
+}
+
+void
+debugMonitorNotify(DebugRawMonitor *dbg_monitor)
+{
+    jrawMonitorID monitor = dbg_monitor->monitor;
+
+    if (!gdata->rankedMonitors) {
+        debugMonitorNotify_norank(monitor);
+        return;
+    }
+
+    jthread current_thread = threadControl_currentThread();
+
+    if (gdata->vmDead && current_thread == NULL) {
+        return;
+    }
+    JNIEnv *env = getEnv();
+    JNI_FUNC_PTR(env,DeleteLocalRef)(env, current_thread);
+
+    // Assert that the current thread owns this monitor.
+    assertIsCurrentThread(getEnv(), dbg_monitor->ownerThread);
```

Comment:
reading ownerThread (and other) fields without dbgRawMonitor is not MT-safe (it's not atomic)

Reply:
It is safe if you are the owner of the monitor.  So what that means is that this assert could potentially crash, but only if you are not the owner, which means if it didn't crash then it would have asserted. However, even when asserts are disabled, isSameObject() is still executed, and that's were the crash would happen. assertIsCurrentThread() really should be a no-op if asserts are not enabled. But even if that change is not made, it still would only crash if something bad was going on since the current thread should always own the monitor.

Reply:
I fixed the issue with executing assertIsCurrentThread even when asserts are disabled.